File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/notifications/NotificationViewModel.kt
package com.sukashawarma.customer.ui.notifications

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sukashawarma.customer.data.NotificationStore
import com.sukashawarma.customer.data.Repository
import com.sukashawarma.customer.data.api.GatewayError
import com.sukashawarma.customer.data.api.GatewayResult
import com.sukashawarma.customer.data.api.NotificationDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class KategoriNotifikasi(val label: String) {
    SEMUA("Semua"),
    PESANAN("Pesanan"),
    PROMO("Promo & Info")
}

data class NotificationState(
    val memuat: Boolean = true,
    val galat: GatewayError? = null,
    val semuaNotifikasi: List<NotificationDto> = emptyList(),
    val unreadCount: Int = 0,
    val tabTerpilih: KategoriNotifikasi = KategoriNotifikasi.SEMUA,
    val sedangMemproses: Boolean = false
) {
    val notifikasiTampil: List<NotificationDto>
        get() = when (tabTerpilih) {
            KategoriNotifikasi.SEMUA -> semuaNotifikasi
            KategoriNotifikasi.PESANAN -> semuaNotifikasi.filter {
                it.type == "order_status" || it.type == "reminder"
            }
            KategoriNotifikasi.PROMO -> semuaNotifikasi.filter {
                it.type == "promo" || it.type == "system"
            }
        }
}

class NotificationViewModel(
    private val repository: Repository,
    private val notificationStore: NotificationStore
) : ViewModel() {

    private val _state = MutableStateFlow(NotificationState())
    val state: StateFlow<NotificationState> = _state.asStateFlow()

    init {
        muat()
    }

    fun muat() {
        viewModelScope.launch {
            _state.value = _state.value.copy(memuat = true, galat = null)
            when (val hasil = repository.ambilNotifikasi()) {
                is GatewayResult.Sukses -> {
                    val notifs = hasil.data.notifications
                    val unread = hasil.data.unreadCount
                    notificationStore.setUnreadCount(unread)
                    _state.value = _state.value.copy(
                        memuat = false,
                        semuaNotifikasi = notifs,
                        unreadCount = unread,
                        galat = null
                    )
                }
                is GatewayResult.Gagal -> {
                    _state.value = _state.value.copy(
                        memuat = false,
                        galat = hasil.error
                    )
                }
            }
        }
    }

    fun gantiTab(tab: KategoriNotifikasi) {
        _state.value = _state.value.copy(tabTerpilih = tab)
    }

    fun tandaiDibaca(id: String) {
        viewModelScope.launch {
            val listLama = _state.value.semuaNotifikasi
            val target = listLama.find { it.id == id } ?: return@launch
            if (target.isRead) return@launch

            // Optimistic update
            val listBaru = listLama.map { if (it.id == id) it.copy(isRead = true) else it }
            val unreadBaru = (_state.value.unreadCount - 1).coerceAtLeast(0)
            notificationStore.setUnreadCount(unreadBaru)
            _state.value = _state.value.copy(
                semuaNotifikasi = listBaru,
                unreadCount = unreadBaru
            )

            repository.tandaiNotifikasiDibaca(notificationId = id)
        }
    }

    fun tandaiSemuaDibaca() {
        viewModelScope.launch {
            if (_state.value.unreadCount == 0) return@launch
            _state.value = _state.value.copy(sedangMemproses = true)

            // Optimistic update
            val listBaru = _state.value.semuaNotifikasi.map { it.copy(isRead = true) }
            notificationStore.setUnreadCount(0)
            _state.value = _state.value.copy(
                semuaNotifikasi = listBaru,
                unreadCount = 0,
                sedangMemproses = false
            )

            repository.tandaiNotifikasiDibaca(tandaiSemua = true)
        }
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Notifications/NotificationViewModel.swift
import Foundation
import Observation

/// Padanan `KategoriNotifikasi` (NotificationViewModel.kt).
enum KategoriNotifikasi: CaseIterable {
    case semua, pesanan, promo

    var label: String {
        switch self {
        case .semua: "Semua"
        case .pesanan: "Pesanan"
        case .promo: "Promo & Info"
        }
    }

    func saring(_ n: [NotificationDto]) -> [NotificationDto] {
        switch self {
        case .semua: n
        case .pesanan: n.filter { $0.type == "order_status" || $0.type == "reminder" }
        case .promo: n.filter { $0.type == "promo" || $0.type == "system" }
        }
    }
}

/// Padanan `NotificationViewModel.kt`. Menandai dibaca bersifat optimistis:
/// tampilan berubah dulu, gateway menyusul.
@MainActor
@Observable
final class NotificationViewModel {
    private(set) var memuat = true
    private(set) var galat: GatewayError?
    private(set) var semua: [NotificationDto] = []
    private(set) var unreadCount = 0
    var tab: KategoriNotifikasi = .semua

    var tampil: [NotificationDto] { tab.saring(semua) }

    @ObservationIgnored private let repository: Repository
    @ObservationIgnored private let store: NotificationStore

    init(repository: Repository, store: NotificationStore) {
        self.repository = repository
        self.store = store
    }

    func muat() async {
        memuat = semua.isEmpty
        galat = nil
        switch await repository.ambilNotifikasi() {
        case .sukses(let d):
            semua = d.notifications
            unreadCount = d.unreadCount
            store.setUnreadCount(d.unreadCount)
        case .gagal(let g):
            galat = g
        }
        memuat = false
    }

    func tandaiDibaca(_ id: String) async {
        guard let i = semua.firstIndex(where: { $0.id == id }), !semua[i].isRead else { return }
        semua[i].isRead = true
        unreadCount = max(unreadCount - 1, 0)
        store.setUnreadCount(unreadCount)
        _ = await repository.tandaiNotifikasiDibaca(notificationId: id)
    }

    func tandaiSemuaDibaca() async {
        guard unreadCount > 0 else { return }
        for i in semua.indices { semua[i].isRead = true }
        unreadCount = 0
        store.setUnreadCount(0)
        _ = await repository.tandaiNotifikasiDibaca(tandaiSemua: true)
    }
}
